strawberryfields/engine.py

```diff
@@ -297,7 +299,7 @@ def _broadcast_nones(val, dim):
             prev = p
 
         if self.samples is not None:
-            return Result(self.samples.copy())
+            return Result(np.array(self.samples).T)
```

Comment:
@pauljxtan @thisac How come there is a transpose needed here? Is there still an inconsistent return shape between some parts of the pipeline?

Reply:
I think it's due to the way the samples are computed by the simulator backend (?) I don't understand the local backend well enough to modify the code there without a risk of breaking something, so will have to defer to someone else's expertise here :slightly_smiling_face:

Reply:
`self.samples` has shape `(modes, shots)` and thus needs to be transposed to `(shots, modes)` somewhere. In the master branch the transposition happens in the `Results` class at the line `np.stack(samples, 1)`, so the `BaseEngine._run` function simply returns `self.samples.copy()` (with shape `(modes, shots)`). I'm not sure why this is. It was like that before I touched it; I only added the specific check in `Results` for a single-shot list so that it returned an array of shape `(shots, modes)` even in that case. But this was all done in the master branch, and not the demo branch. Now both branches seem to have different implementations of the `Results` class. I'm not too sure about what the demo branch is though, so maybe this doesn't matter that much. :slightly_smiling_face:

Reply:
Thanks! This is making more sense to me now. What I essentially did was move the transposition slightly down the pipeline, from `Result.__init__` to the end of `BaseEngine._run`. This prevents `np.stack(samples, 1)` (now removed) from incorrectly transposing samples coming in from the remote platform (which have `(shots, modes)`). So I believe this all works out.

Reply:
That makes sense. Thanks! :slightly_smiling_face:  Should this change preferably be done in the master branch as well (or will the demo branch be merged with the master branch at some point)? @josh146 @co9olguy

Reply:
The plan is to merge the demo branch into master in the coming week(s), before the upcoming SF release!

Reply:
However, it sounds like there will be a merge conflict now, as master does the transposition in `Result.__init__`, and demo does it in `BaseEngine._run`, so we will need to be careful to resolve this conflict when we merge.

Reply:
I think we will have to look at this again carefully after the merge. In an ideal world, the API _as well as_ the Gaussian/Fock/TF backends, should be returning samples of the form `(shots, modes)`. The engine or the results class should not have to do any transposition.

Reply:
I looked through how the `LocalEngine` gets samples from the backends, and it doesn't look like there's an easy fix to get the correct shape directly from the backends. The only backend that supports multi-shot measurements seems to be the Gaussian backend, and it returns an array of values for one mode at a time (one value for each shot). In that case the combining of modes is done in `engine.py` either way, so the transposing would have to be there also. Maybe I misunderstood something though.

As for the other backends they don't support multiple shots together with any measurements, and simply return a list of length = number of modes filled with `None`, which is then converted into a nested array in the desired shape in `engine.py` (the `_broadcast_nones` function in `BaseEngine._run`).